File: problems/2022_09_17/0543/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int diameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return diameter;
    }

    private int helper(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int leftDiameter = helper(root.left);
        int rightDiameter = helper(root.right);

        diameter = Math.max(diameter, leftDiameter + rightDiameter);

        return Math.max(leftDiameter, rightDiameter) + 1;
    }
}

File: problems/2022_09_17/0543/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.longest = -1
        self.getLength(root)
        return self.longest

    def getLength(self, root):
        if not root:
            return -1

        left = self.getLength(root.left)
        right = self.getLength(root.right)

        self.longest = max(self.longest, 2 + left + right)

        return 1 + max(left, right)
